core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -492,13 +495,152 @@ public void failTaskRemotely(TaskId taskId, Throwable failureCause)
             SqlStage sqlStage = requireNonNull(stages.get(taskId.getStageId()), () -> "stage not found: %s" + taskId.getStageId());
             sqlStage.failTaskRemotely(taskId, failureCause);
         }
+
+        public void logDebugInfo()
+        {
+            if (!log.isDebugEnabled()) {
+                return;
+            }
+            log.debug("SqlStages:");
+            stages.forEach((stageId, stage) -> {
+                log.debug("SqlStage %s: %s", stageId, stage);
+            });
+        }
+    }
+
+    private static class EventDebugInfos
+    {
+        private static final String GLOBAL_EVENTS_BUCKET = "GLOBAL";
+        private static final EventListener<String> GET_BUCKET_LISTENER = new EventListener<>()
+        {
+            @Override
+            public String onRemoteTaskEvent(RemoteTaskEvent event)
+            {
+                return "task_" + event.getTaskStatus().getTaskId().getStageId().toString();
+            }
+
+            @Override
+            public String onRemoteTaskExchangeUpdatedSinkAcquired(RemoteTaskExchangeUpdatedSinkAcquired event)
+            {
+                return "task_" + event.getTaskId().getStageId().toString();
+            }
+
+            @Override
+            public String onStageEvent(StageEvent event)
+            {
+                return event.getStageId().toString();
+            }
+
+            @Override
+            public String onEvent(Event event)
+            {
+                return GLOBAL_EVENTS_BUCKET;
+            }
+        };
+
+        private final String queryId;
+        private final int eventsPerBucket;
+        private long eventsCounter;
+        private long filteredEventsCounter;
+
+        private volatile SoftReference<ListMultimap<String, String>> eventsDebugInfosReference;
+
+        private EventDebugInfos(String queryId, int eventsPerBucket)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.eventsPerBucket = eventsPerBucket;
+            eventsDebugInfosReference = new SoftReference<>(LinkedListMultimap.create());
+        }
+
+        private boolean add(Event event)
+        {
+            ListMultimap<String, String> eventsDebugInfos = getEventsDebugInfos();
+            String bucket = getBucket(event);
+            Optional<String> debugInfo = getFullDebugInfo(eventsCounter, event);
+            eventsCounter++;
+            if (debugInfo.isEmpty()) {
+                filteredEventsCounter++;
+                return false;
+            }
+            eventsDebugInfos.put(bucket, debugInfo.get());
+            List<String> bucketDebugInfos = eventsDebugInfos.get(bucket);
+            if (bucketDebugInfos.size() > eventsPerBucket) {
```

Comment:
since you operate on `bucketDebugInfos` directly, 
using `eventsDebugInfos.put` doesn't buy any convenience
in fact, it blurs the guarantee of size enforcement (size is enforced on a different collection than being added to)